pkg/tunneler/listener.go

```diff
@@ -0,0 +1,290 @@
+	j = append(j, '\n')
+	ln.writec <- j
+}
+
+func (ln *Listener) dial() (net.Conn, error) {
+	pr, pw := io.Pipe()
+	req, err := http.NewRequest("GET", ln.url, pr)
+	if err != nil {
+		klog.V(5).Infof("Can not create request %v", err)
+		return nil, err
+	}
+
+	klog.V(5).Infof("Listener creating connection to %s", ln.url)
+	res, err := ln.client.Do(req)
+	if err != nil {
+		klog.V(5).Infof("Can not connect to %s request %v, retry %d", ln.url, err)
+		return nil, err
+	}
+	if res.StatusCode != 200 {
+		klog.V(5).Infof("Status code %d on request %v, retry %d", res.StatusCode, ln.url)
+		return nil, fmt.Errorf("status code %d", res.StatusCode)
+	}
+
+	conn := rwconn.NewConn(res.Body, pw)
+	return conn, nil
+}
+
+func (ln *Listener) grabConn() {
+	// create a new connection
+	c, err := ln.dial()
+	if err != nil {
+		klog.V(5).Infof("Can not create connection %v", err)
+		ln.sendMessage(controlMsg{Command: "pickup-failed", ConnPath: "", Err: err.Error()})
+		return
+	}
+
+	// send the connection to the listener
+	select {
+	case <-ln.donec:
+		return
+	default:
+		select {
+		case ln.connc <- c:
+		case <-ln.donec:
+			return
+		}
+	}
+
+}
+
+// Accept blocks and returns a new connection, or an error.
+func (ln *Listener) Accept() (net.Conn, error) {
+	c, ok := <-ln.connc
+	if !ok {
+		ln.mu.Lock()
+		err, closed := ln.readErr, ln.closed
+		ln.mu.Unlock()
+		if err != nil && !closed {
+			return nil, fmt.Errorf("tunneler: Listener closed; %w", err)
+		}
+		return nil, ErrListenerClosed
+	}
+	klog.V(5).Infof("Accept connection")
+	return c, nil
+}
+
+// ErrListenerClosed is returned by Accept after Close has been called.
+var ErrListenerClosed = errors.New("tunneler: Listener closed")
+
+// Close closes the Listener, making future Accept calls return an
+// error.
+func (ln *Listener) Close() error {
+	ln.mu.Lock()
+	defer ln.mu.Unlock()
+	if ln.closed {
+		return nil
+	}
+	ln.closed = true
+	close(ln.connc)
+	close(ln.donec)
+	ln.sc.Close()
+	return nil
+}
+
+// Addr returns a dummy address. This exists only to conform to the
+// net.Listener interface.
+func (ln *Listener) Addr() net.Addr { return connAddr{} }
+
+// configureHTTP2Transport enable ping to avoid issues with stale connections
+func configureHTTP2Transport(client *http.Client) error {
+	t, ok := client.Transport.(*http.Transport)
+	if !ok {
+		// can't get the transport it will fail later if not http2 supported
+		return nil
+	}
+
+	if t.TLSClientConfig == nil {
+		return fmt.Errorf("only TLS supported")
+	}
+
+	for _, v := range t.TLSClientConfig.NextProtos {
+		// http2 already configured
+		if v == "h2" {
+			return nil
+		}
+	}
+
+	t2, err := http2.ConfigureTransports(t)
+	if err != nil {
+		return err
+	}
+
+	t2.ReadIdleTimeout = time.Duration(30) * time.Second
+	t2.PingTimeout = time.Duration(15) * time.Second
+	return nil
+}
+
+// syncerTunnelURL builds the destination url with the Dialer expected format of the URL
+func syncerTunnelURL(host string, id string) (string, error) {
```

Comment:
where does the forked tunneler.go code end? Can we move everything into a different file that is not coming from the upstream file?

Reply:
the code is not really forked, there are some tweaks I've done